doc/move_group_interface/src/move_group_interface_tutorial.cpp

```diff
@@ -1,394 +1,648 @@
+
+    // We want the Cartesian path to be interpolated at a resolution of 1 cm
+    // which is why we will specify 0.01 as the max step in Cartesian
+    // translation.  We will specify the jump threshold as 0.0, effectively disabling it.
+    // Warning - disabling the jump threshold while operating real hardware can cause
+    // large unpredictable motions of redundant joints and could be a safety issue.
+    moveit_msgs::RobotTrajectory trajectory;
+    const double jump_threshold = 0.0;
+    const double eef_step = 0.01;
+    double fraction = move_group.computeCartesianPath(waypoints, eef_step, jump_threshold, trajectory);
+    ROS_INFO_NAMED("tutorial", "Visualizing plan 5 (Cartesian path) (%.2f%% acheived)", fraction * 100.0);
+
+    // While we didn't use the planner to create this trajectory it still can be added to a plan
+    my_plan.trajectory_ = trajectory;
+
+    // END_SUB_TUTORIAL
+
+    // Visualize the plan in RViz, becuase there are multiple points we use a loop to show them all
+    visual_tools.publishPath(waypoints, rvt::LIME_GREEN, rvt::SMALL);
+    for (std::size_t i = 0; i < waypoints.size(); ++i)
+      visual_tools.publishAxisLabeled(waypoints[i], "pt" + std::to_string(i), rvt::SMALL);
+    visual_tools.trigger();
+    visual_tools.prompt("Press 'next' to move along the cartesian path");
+  }
+
+  void create_obstacle()
+  {
+    // BEGIN_SUB_TUTORIAL add_obj
+
+    // Remember to change the velocity scaling factor back after moving to return to full speed.
+    move_group.setMaxVelocityScalingFactor(1);
+
+    // Adding or Removing Objects
+    // ^^^^^^^^^^^^^^^^^^^^^^^^^^
+    //
+    // First Define a collision object ROS message.
+    moveit_msgs::CollisionObject collision_object;
+    collision_object.header.frame_id = move_group.getPlanningFrame();
+
+    // Then set the id of the object is used to identify it.
+    collision_object.id = "box1";
+
+    // Now we can Define a box to add to the world.
+    shape_msgs::SolidPrimitive primitive;
+    primitive.type = primitive.BOX;
+    primitive.dimensions.resize(3);
+    primitive.dimensions[0] = 0.02;
+    primitive.dimensions[1] = 0.4;
+    primitive.dimensions[2] = 0.4;
+
+    // Define a pose for the box. (specified relative to frame_id)
+    geometry_msgs::Pose box_pose;
+    box_pose.orientation.w = 1.0;
+    box_pose.position.x = 0.3;
+    box_pose.position.y = -0.25;
+    box_pose.position.z = 0.5;
+
+    collision_object.primitives.push_back(primitive);
+    collision_object.primitive_poses.push_back(box_pose);
+    collision_object.operation = collision_object.ADD;
+
+    collision_objects.push_back(collision_object);
+    object_ids.push_back(collision_object.id);
+
+    // Now, let's add the collision object into the world.
+    planning_scene_interface.addCollisionObjects(collision_objects);
+    // END_SUB_TUTORIAL
+
+    // RViz
+    ROS_INFO_NAMED("tutorial", "Add an object into the world");
+    visual_tools.deleteAllMarkers();
```

Comment:
If you find yourself copy-pasting code a number of times, it is a good indication that you should put that functionality into a separate method. In this case,`publishStepAndWait(const std::string& step, const std::string& prompt_text);` would be one way to refactor this out.